rebis-dev: Fix Heap::compute_pstr_size being off by one cell

adri326 opened the pull request:
So it looks like my fix for segfault around heap and PStr had an off-by-one error in the logic for `compute_pstr_size`, which this PR fixes.
This off-by-one error was compounded by a second off-by-one error within the logic of `build_functor!()` (some parts did not account for the `[]` atom appended to `PStr`s), which happened to cancel out with the previous, invalid implementation of `compute_pstr_size`.

This does not address #2757 (and I would prefer to fix that issue in a separate PR, so that we can get the unit tests of `rebis-dev` back in order), though the logic for `compute_pstr_size` *should* coincide with the corrected behavior of #2757. The test cases I've added here ought to be tweaked once the correct behavior is set in stone.

Now `Heap::compute_pstr_size` accurately measures the number of bytes needed by `Heap::allocate_pstr` (except in the instances of #2757).
I've also simplified the logic a bit, and cleaned up the affected tests of `build_functor` :)

triska commented on src/machine/heap.rs:
```diff
@@ -1044,31 +1041,23 @@ impl Heap {
     /// Returns the number of bytes needed to store `src` as a `PStr`.
     pub(crate) fn compute_pstr_size(src: &str) -> usize {
+        if src.is_empty() {
+            // As of now, null strings aren't allocated.
```
The string `""` is the atom `[]`, and that atom should be used to represent the string.

<pre>
<b>?- "" == [].</b>
   true.
</pre>

adri326 replied:
I've amended the comment to now say:

> Empty strings are represented by the '[]' atom.
> As such, no PStr is needed or allocated:

adri326 commented:
It turns out that I had missed one instance of the issue spotted in #2729: `scan_slice_to_str` wrongly assumed that `Heap.inner.ptr` was aligned to `Heap::heap_cell_alignment()`.

I've rewritten it (and renamed it to the more accurate `scan_pstr_at`) to not rely on this incorrect assumption anymore, and to reduce the number of `unsafe` code needed in it and around it, by expressing it in terms of `HeapView` instead.

To make it easy to use it in the implementations of `SizedHeap`, I've introduced a small `AsHeapView` trait, currently only local to this module.

triska commented:
A few CI tests apparently fail, I hope at least the formatting would be easy to resolve.

adri326 commented:
The formatting issue is omnipresent on rebis-dev, running `carfo fmt` gives me a few thousands of lines of diff (I made that mistake multiple times). I would prefer if @mthom pushed a `cargo fmt` commit on rebis-dev and I can then work through the rebase of my branches

triska commented:
Does 823b74c18a1254d17f9b8941057bc4bd3adbb918 address the same issue as eb16e54daaecff0db3de80a835e3a29b8faaaac0?

mthom commented:
@adri326 Just pushed rebis-dev with a cargo fmt commit.

triska commented:
@adri326: If you have time, could you please rebase the commits on top of the current `rebis-dev` branch? Thank you a lot!

adri326 commented:
The original issue has been fixed by 2cf3b5757 in the meantime
